# EXPLORING NASA’S TURBOFAN DATASET

In [1]:
# libraries
import numpy as np
import pandas as pd 
import matplotlib.pyplot as plt
%matplotlib inline 

Datasets include simulations of multiple turbofan engines over time, each row contains the following information:
1. Engine unit number
2. Time, in cycles
3. Three operational settings
4. 21 sensor readings

In [2]:
# define file path to dataset
filepath = './CMaps/'

# setting column names 
index_names = ['unit_nr', 'time_cycles']
settings = ['setting_1', 'setting_2', 'setting_3']
sensor_names = ['s_{}'.format(i) for i in range(1, 22)]
col_names =  index_names + settings + sensor_names

# load data
train_data = pd.read_csv((filepath + 'train_FD001.txt'), sep='\s+', header=None, names=col_names)
test_data = pd.read_csv((filepath + 'test_FD001.txt'), sep='\s+', header=None, names=col_names)
y_test = pd.read_csv((filepath + 'RUL_FD001.txt'), sep='\s+', header=None, names=['RUL'])

# first rows
train_data.head()

unit_nr  time_cycles  setting_1  setting_2  setting_3     s_1     s_2  \
0        1            1    -0.0007    -0.0004      100.0  518.67  641.82   
1        1            2     0.0019    -0.0003      100.0  518.67  642.15   
2        1            3    -0.0043     0.0003      100.0  518.67  642.35   
3        1            4     0.0007     0.0000      100.0  518.67  642.35   
4        1            5    -0.0019    -0.0002      100.0  518.67  642.37   

       s_3      s_4    s_5  ...    s_12     s_13     s_14    s_15  s_16  s_17  \
0  1589.70  1400.60  14.62  ...  521.66  2388.02  8138.62  8.4195  0.03   392   
1  1591.82  1403.14  14.62  ...  522.28  2388.07  8131.49  8.4318  0.03   392   
2  1587.99  1404.20  14.62  ...  522.42  2388.03  8133.23  8.4178  0.03   390   
3  1582.79  1401.87  14.62  ...  522.86  2388.08  8133.83  8.3682  0.03   392   
4  1582.85  1406.22  14.62  ...  522.19  2388.04  8133.80  8.4294  0.03   393   

   s_18   s_19   s_20     s_21  
0  2388  100.0  39.06  23.4190  
1  2388  100.0  39.00  23.4236  
2  2388  100.0  38.95  23.3442  
3  2388  100.0  38.88  23.3739  
4  2388  100.0  38.90  23.4044  

[5 rows x 26 columns]

In [13]:
train_data[index_names[0]].to_frame().describe()

unit_nr
count  20631.000000
mean      51.506568
std       29.227633
min        1.000000
25%       26.000000
50%       52.000000
75%       77.000000
max      100.000000

In [17]:
train_data.groupby('unit_nr').max().describe()

time_cycles   setting_1   setting_2  setting_3           s_1  \
count   100.000000  100.000000  100.000000      100.0  1.000000e+02   
mean    206.310000    0.005986    0.000497      100.0  5.186700e+02   
std      46.342749    0.000883    0.000056        0.0  1.142596e-12   
min     128.000000    0.004000    0.000400      100.0  5.186700e+02   
25%     177.000000    0.005375    0.000500      100.0  5.186700e+02   
50%     199.000000    0.005900    0.000500      100.0  5.186700e+02   
75%     229.250000    0.006425    0.000500      100.0  5.186700e+02   
max     362.000000    0.008700    0.000600      100.0  5.186700e+02   

              s_2          s_3          s_4           s_5           s_6  ...  \
count  100.000000   100.000000   100.000000  1.000000e+02  1.000000e+02  ...   
mean   644.108800  1608.073100  1434.506400  1.462000e+01  2.161000e+01  ...   
std      0.141773     2.446053     2.323773  2.856489e-14  3.570612e-14  ...   
min    643.870000  1604.070000  1429.260000  1.462000e+01  2.161000e+01  ...   
25%    644.010000  1606.362500  1432.647500  1.462000e+01  2.161000e+01  ...   
50%    644.090000  1607.685000  1434.230000  1.462000e+01  2.161000e+01  ...   
75%    644.190000  1609.350000  1436.275000  1.462000e+01  2.161000e+01  ...   
max    644.530000  1616.910000  1441.490000  1.462000e+01  2.161000e+01  ...   

             s_12         s_13         s_14        s_15          s_16  \
count  100.000000   100.000000   100.000000  100.000000  1.000000e+02   
mean   522.593200  2388.285100  8179.481200    8.547498  3.000000e-02   
std      0.422197     0.073368    43.595736    0.012937  4.881696e-17   
min    521.770000  2388.170000  8128.370000    8.517100  3.000000e-02   
25%    522.237500  2388.230000  8147.500000    8.537525  3.000000e-02   
50%    522.580000  2388.280000  8157.700000    8.548250  3.000000e-02   
75%    523.000000  2388.330000  8215.420000    8.556250  3.000000e-02   
max    523.380000  2388.560000  8293.720000    8.584800  3.000000e-02   

             s_17    s_18   s_19        s_20        s_21  
count  100.000000   100.0  100.0  100.000000  100.000000  
mean   397.580000  2388.0  100.0   39.167800   23.498315  
std      0.684312     0.0    0.0    0.091625    0.057682  
min    396.000000  2388.0  100.0   38.970000   23.375000  
25%    397.000000  2388.0  100.0   39.100000   23.454375  
50%    397.500000  2388.0  100.0   39.155000   23.498900  
75%    398.000000  2388.0  100.0   39.232500   23.543200  
max    400.000000  2388.0  100.0   39.430000   23.618400  

[8 rows x 25 columns]

When we inspect the `unit_nr` descriptive statistics it can be seen that the dataset has a total of 20631 rows, unit members starting at 1 to 100 as expected. From the `time_cycles`, it took some engines a maximum of *362 time cycles* to run to failure and the average engine breaks between *199* and *206 cycles*. However, a standard deviation of *46 cycles* is very large.  

In [19]:
# details of 3 operational settings 
train_data[settings].describe()

setting_1     setting_2  setting_3
count  20631.000000  20631.000000    20631.0
mean      -0.000009      0.000002      100.0
std        0.002187      0.000293        0.0
min       -0.008700     -0.000600      100.0
25%       -0.001500     -0.000200      100.0
50%        0.000000      0.000000      100.0
75%        0.001500      0.000300      100.0
max        0.008700      0.000600      100.0

Looking at the three operating conditions the it can be seen that the standard deviation of both `settings_1` and `settings_2` have small fluctuations but `settings_3` has 0 with other statistics at 100. 

In [22]:
# details of sensors
train_data[sensor_names].describe().transpose()

count         mean           std        min        25%        50%  \
s_1   20631.0   518.670000  6.537152e-11   518.6700   518.6700   518.6700   
s_2   20631.0   642.680934  5.000533e-01   641.2100   642.3250   642.6400   
s_3   20631.0  1590.523119  6.131150e+00  1571.0400  1586.2600  1590.1000   
s_4   20631.0  1408.933782  9.000605e+00  1382.2500  1402.3600  1408.0400   
s_5   20631.0    14.620000  3.394700e-12    14.6200    14.6200    14.6200   
s_6   20631.0    21.609803  1.388985e-03    21.6000    21.6100    21.6100   
s_7   20631.0   553.367711  8.850923e-01   549.8500   552.8100   553.4400   
s_8   20631.0  2388.096652  7.098548e-02  2387.9000  2388.0500  2388.0900   
s_9   20631.0  9065.242941  2.208288e+01  9021.7300  9053.1000  9060.6600   
s_10  20631.0     1.300000  4.660829e-13     1.3000     1.3000     1.3000   
s_11  20631.0    47.541168  2.670874e-01    46.8500    47.3500    47.5100   
s_12  20631.0   521.413470  7.375534e-01   518.6900   520.9600   521.4800   
s_13  20631.0  2388.096152  7.191892e-02  2387.8800  2388.0400  2388.0900   
s_14  20631.0  8143.752722  1.907618e+01  8099.9400  8133.2450  8140.5400   
s_15  20631.0     8.442146  3.750504e-02     8.3249     8.4149     8.4389   
s_16  20631.0     0.030000  1.556432e-14     0.0300     0.0300     0.0300   
s_17  20631.0   393.210654  1.548763e+00   388.0000   392.0000   393.0000   
s_18  20631.0  2388.000000  0.000000e+00  2388.0000  2388.0000  2388.0000   
s_19  20631.0   100.000000  0.000000e+00   100.0000   100.0000   100.0000   
s_20  20631.0    38.816271  1.807464e-01    38.1400    38.7000    38.8300   
s_21  20631.0    23.289705  1.082509e-01    22.8942    23.2218    23.2979   

            75%        max  
s_1    518.6700   518.6700  
s_2    643.0000   644.5300  
s_3   1594.3800  1616.9100  
s_4   1414.5550  1441.4900  
s_5     14.6200    14.6200  
s_6     21.6100    21.6100  
s_7    554.0100   556.0600  
s_8   2388.1400  2388.5600  
s_9   9069.4200  9244.5900  
s_10     1.3000     1.3000  
s_11    47.7000    48.5300  
s_12   521.9500   523.3800  
s_13  2388.1400  2388.5600  
s_14  8148.3100  8293.7200  
s_15     8.4656     8.5848  
s_16     0.0300     0.0300  
s_17   394.0000   400.0000  
s_18  2388.0000  2388.0000  
s_19   100.0000   100.0000  
s_20    38.9500    39.4300  
s_21    23.3668    23.6184

It can be observed that the standard deviations of sensor 18, and 19 have 0 fluctuations nad these can safely be discarded as they hild no useful information. Inspecting the quantiles indicates sensor 1, 5, 6, 10 and 15 have very little fluctuations and requires further inspection while sensor 9 and 14 seems to have the highest fluctuation.

## Computing RUL

Remaining Useful Life (RUL) is the target variable which will be used to serve two purposes:
1. It will be used to plot against the signals for easy interpretation of the changes in each sensor signals as the engine nears breakdown
2. It will serve as the target variable for the supervised machine learning models

`RUL = max_time_cycle - time_cycle`

In [23]:
def add_remaining_useful_life(df):
    # Get total number of cycles for each unit
    grouped_by_unit = df.groupby('unit_nr')
    max_cycle = grouped_by_unit['time_cycles'].max()
    
    # Merge the max cycle back into the original dataframe
    result_frame = df.merge(max_cycle.to_frame(name='max_cycle'), left_on='unit_nr',
                           right_index=True)
    
    # Calculate remaining useful life for each row
    remaining_useful_life = result_frame['max_cycle'] - result_frame['time_cycles']
    result_frame['RUL'] = remaining_useful_life
    
    # drop max_cycle 
    result_frame = result_frame.drop('max_cycle', axis=1)
    return result_frame

train_data = add_remaining_useful_life(train_data)
train_data[index_names + ['RUL']].head()

unit_nr  time_cycles  RUL
0        1            1  191
1        1            2  190
2        1            3  189
3        1            4  188
4        1            5  187